pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -3543,98 +3543,88 @@ protected void internalExpireMessagesByPosition(AsyncResponse asyncResponse, Str
+                            if (exception != null) {
+                                Throwable th = exception.getCause();
+                                if (th instanceof NotFoundException) {
+                                    asyncResponse.resume(new RestException(Status.NOT_FOUND, th.getMessage()));
+                                    return null;
+                                } else if (th instanceof WebApplicationException) {
+                                    asyncResponse.resume(th);
+                                    return null;
+                                } else {
+                                    log.error("[{}] Failed to trigger compaction on topic {}",
+                                            clientAppId(), topicName, exception);
+                                    asyncResponse.resume(new RestException(exception));
+                                    return null;
+                                }
                             }
+                            asyncResponse.resume(Response.noContent().build());
+                            return null;
+                        });
+                    } else {
+                        try {
+                            internalTriggerCompactionNonPartitionedTopic(authoritative);
+                        } catch (Exception e) {
+                            if (e instanceof WebApplicationException) {
+                                if (log.isDebugEnabled()) {
+                                    log.debug("[{}] Failed to trigger compaction on topic {}, "
+                                            + "redirecting to other brokers.", clientAppId(), topicName, e);
+                                }
+                                resumeAsyncResponseExceptionally(asyncResponse, e);
+                                return;
+                            }
+                            log.error("[{}] Failed to trigger compaction on topic {}", clientAppId(), topicName, e);
                             resumeAsyncResponseExceptionally(asyncResponse, e);
                             return;
                         }
-                        log.error("[{}] Failed to trigger compaction on topic {}", clientAppId(), topicName, e);
-                        resumeAsyncResponseExceptionally(asyncResponse, e);
-                        return;
+                        asyncResponse.resume(Response.noContent().build());
                     }
-                    asyncResponse.resume(Response.noContent().build());
-                }
-            }).exceptionally(ex -> {
-                log.error("[{}] Failed to trigger compaction on topic {}", clientAppId(), topicName, ex);
-                resumeAsyncResponseExceptionally(asyncResponse, ex);
-                return null;
-            });
-        }
+                }).exceptionally(ex -> {
+                    log.error("[{}] Failed to trigger compaction on topic {}", clientAppId(), topicName, ex);
```

Comment:
``Failed to validate global namespace ownership to trigger compaction on topic``